beacon_node/operation_pool/src/lib.rs

```diff
@@ -235,31 +236,40 @@ impl<T: EthSpec> OperationPool<T> {
         (proposer_slashings, attester_slashings)
     }
 
-    /// Prune proposer slashings for all slashed or withdrawn validators.
-    pub fn prune_proposer_slashings(&self, finalized_state: &BeaconState<T>) {
+    /// Prune proposer slashings for validators which are exited in the finalized epoch.
+    pub fn prune_proposer_slashings(&self, head_state: &BeaconState<T>) {
         prune_validator_hash_map(
             &mut self.proposer_slashings.write(),
-            |validator| {
-                validator.slashed || validator.is_withdrawable_at(finalized_state.current_epoch())
-            },
-            finalized_state,
+            |validator| validator.exit_epoch <= head_state.finalized_checkpoint.epoch,
+            head_state,
         );
     }
 
     /// Prune attester slashings for all slashed or withdrawn validators, or attestations on another
     /// fork.
-    pub fn prune_attester_slashings(&self, finalized_state: &BeaconState<T>, head_fork: Fork) {
+    pub fn prune_attester_slashings(&self, head_state: &BeaconState<T>) {
         self.attester_slashings
             .write()
             .retain(|(slashing, fork_version)| {
-                // Any slashings for forks older than the finalized state's previous fork can be
-                // discarded. We allow the head_fork's current version too in case a fork has
-                // occurred between the finalized state and the head.
-                let fork_ok = *fork_version == finalized_state.fork.previous_version
-                    || *fork_version == finalized_state.fork.current_version
-                    || *fork_version == head_fork.current_version;
+                let previous_fork_is_finalized =
+                    head_state.finalized_checkpoint.epoch >= head_state.fork.epoch;
+                // Prune any slashings which don't match the current fork version, or the previous
+                // fork version if it is not finalized yet.
+                let fork_ok = (fork_version == &head_state.fork.current_version)
+                    || (fork_version == &head_state.fork.previous_version
+                        && !previous_fork_is_finalized);
                 // Slashings that don't slash any validators can also be dropped.
-                let slashing_ok = get_slashable_indices(finalized_state, slashing).is_ok();
+                let slashing_ok =
+                    get_slashable_indices_modular(head_state, slashing, |_, validator| {
+                        // Declare that a validator is still slashable if they have not exited prior
+                        // to the finalized epoch.
+                        //
+                        // We cannot check the `slashed` field since the `head` is not finalized and
+                        // a fork could un-slash someone.
+                        validator.exit_epoch > head_state.finalized_checkpoint.epoch
+                    })
+                    .map_or(false, |indices| !indices.is_empty());
```

Comment:
The `!indices.is_empty()` check is redundant because that's already checked by `get_slashable_indices*`, here:

https://github.com/sigp/lighthouse/blob/a60ab4eff27e1f34e6c4d700d68ad7029985de14/consensus/state_processing/src/per_block_processing/verify_attester_slashing.rs#L95

Reply:
I saw that, I just felt I should be a little more defensive in case we lose that check in the function.

Reply:
Fair enough, I'm happy to keep it